pkg/component/extensions/operatingsystemconfig/operatingsystemconfig.go

```diff
@@ -220,23 +230,174 @@ func (o *operatingSystemConfig) reconcile(ctx context.Context, reconcileFn func(
 		return err
 	}
 
-	fns := o.forEachWorkerPoolAndPurposeTaskFn(func(_ context.Context, osc *extensionsv1alpha1.OperatingSystemConfig, worker gardencorev1beta1.Worker, purpose extensionsv1alpha1.OperatingSystemConfigPurpose) error {
-		d, err := o.newDeployer(osc, worker, purpose)
+	if err := o.updateVersioningSecret(ctx); err != nil {
+		return err
+	}
+
+	fns, err := o.forEachWorkerPoolAndPurposeTaskFn(func(_ context.Context, version int, osc *extensionsv1alpha1.OperatingSystemConfig, worker gardencorev1beta1.Worker, purpose extensionsv1alpha1.OperatingSystemConfigPurpose) error {
+		d, err := o.newDeployer(version, osc, worker, purpose)
 		if err != nil {
 			return err
 		}
 
 		return reconcileFn(d)
 	})
+	if err != nil {
+		return err
+	}
 
 	return flow.Parallel(fns...)(ctx)
 }
 
+type poolHash struct {
+	Pools []poolHashEntry
+}
+
+type poolHashEntry struct {
+	Name           string
+	CurrentVersion int
+	Hashes         map[int]string
```

Comment:
```suggestion
	HashVersionToPoolName         map[int]string
```

Reply:
I've renamed it to `HashVersionToOSCKey`. `PoolName` is isn't a good fit IMO.